llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -1820,35 +1820,61 @@ Instruction *InstCombinerImpl::foldOpIntoPhi(Instruction &I, PHINode *PN) {
       continue;
     }
 
-    if (NonSimplifiedBB) return nullptr;  // More than one non-simplified value.
+    // If the only use of phi is comparing it with a constant then we can
+    // put this comparison in the incoming BB directly after a ucmp/scmp call
+    // because we know that it will simplify to a single icmp.
+    const APInt *Ignored;
+    if (isa<CmpIntrinsic>(InVal) && InVal->hasOneUse() &&
+        match(&I, m_c_ICmp(m_Specific(PN), m_APInt(Ignored)))) {
```

Comment:
```suggestion
        match(&I, m_ICmp(m_Specific(PN), m_APInt(Ignored)))) {
```
We always place constants in RHS.